transport/nio/src/test/java/org/kaazing/gateway/transport/tcp/TcpConnectorIT.java

```diff
@@ -0,0 +1,203 @@
+    public void serverSentData() throws Exception {
+        bindTo8080(new IoHandlerAdapter(){
+            @Override
+            protected void doSessionOpened(IoSession session) throws Exception {
+             // KG-8210: push pending write requests onto the write request queue before disposing, should not cause hang in dispose
+                ByteBuffer data = ByteBuffer.allocate(20);
+                String str = "server data";
+                data.put(str.getBytes());
+                
+                data.flip();
+
+                IoBufferAllocatorEx<?> allocator = ((IoSessionEx) session).getBufferAllocator();
+                
+                session.write(allocator.wrap(data.duplicate(), IoBufferEx.FLAG_SHARED));
+            }
+        });
+        
+        k3po.finish();
+    }
+
+    @Test
+    @Specification({
+        "client.sent.data/tcp.client"
+        })
+    public void clientSentData() throws Exception {
+        bindTo8080(new IoHandlerAdapter());
+        k3po.finish();
+    }
+
+    @Test
+    @Specification({
+        "bidirectional.data/tcp.client"
+        })
+    public void bidirectionalData() throws Exception {
+        bindTo8080(new IoHandlerAdapter(){
+            private int counter = 1;
+            @Override
+            protected void doMessageReceived(IoSession session, Object message) throws Exception {
+             // KG-8210: push pending write requests onto the write request queue before disposing, should not cause hang in dispose
+                ByteBuffer data = ByteBuffer.allocate(20);
+                String str = "server data " + counter;
+                counter++;
+                data.put(str.getBytes());
+                
+                data.flip();
+
+                IoBufferAllocatorEx<?> allocator = ((IoSessionEx) session).getBufferAllocator();
+                
+                session.write(allocator.wrap(data.duplicate(), IoBufferEx.FLAG_SHARED));
+            }
+        });
+        
+        
+        k3po.finish();
+    }
+
+    @Test
+    @Specification({
+        "server.close/tcp.client"
+        })
+    public void serverClose() throws Exception {
+        bindTo8080(new IoHandlerAdapter(){
+            @Override
+            protected void doSessionOpened(IoSession session) throws Exception {
+                session.close();
+            }
+        });
+        
+        k3po.finish();
+    }
+
+    @Test
+    @Specification({
+        "client.close/tcp.client"
+        })
+    public void clientClose() throws Exception {
+        bindTo8080(new IoHandlerAdapter());
+        k3po.finish();
+    }
+
+    @Test
+    @Specification({
+        "concurrent.connections/tcp.client"
+        })
+    public void concurrentConnections() throws Exception {
+        bindTo8080(new IoHandlerAdapter(){
+            @Override
+            protected void doMessageReceived(IoSession session, Object message) throws Exception {                
+                session.write(message);
```

Comment:
Need to read and check the incoming data, allowing for fragmentation, before writing